cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,301 @@
+		}
+
+	case Latest:
+	default:
+		return errors.New("wrong recover type")
+	}
+
+	for _, binlog := range r.binlogs {
+		log.Println("working with", binlog)
+
+		if r.recoverType == Date {
+			binlogArr := strings.Split(binlog, "_")
+			if len(binlogArr) < 2 {
+				return errors.New("get timestamp from binlog name")
+			}
+			binlogTime, err := strconv.ParseInt(binlogArr[1], 10, 64)
+			if err != nil {
+				return errors.Wrap(err, "get binlog time")
+			}
+			if binlogTime > endTime.Unix() {
+				return nil
+			}
+		}
+
+		binlogObj, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get obj")
+		}
+
+		err = os.Setenv("MYSQL_PWD", os.Getenv("PXC_PASS"))
+		if err != nil {
+			return errors.Wrap(err, "set mysql pwd env var")
+		}
+
+		cmdString := "mysqlbinlog" + flags + " - | mysql -h" + r.db.GetHost() + " -u" + r.pxcUser
+		cmd := exec.Command("sh", "-c", cmdString)
+
+		cmd.Stdin = binlogObj
+		var outb, errb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Stderr = &errb
+		err = cmd.Run()
+		if err != nil {
+			return errors.Wrapf(err, "cmd run. stderr: %s, stdout: %s", errb.String(), outb.String())
+		}
+	}
+
+	return nil
+}
+
+func getLastBackupGTID(infoObj io.Reader) (string, error) {
+	sep := []byte("GTID of the last")
+
+	content, err := ioutil.ReadAll(infoObj)
+	if err != nil {
+		return "", errors.Wrap(err, "read object")
+	}
+
+	startIndex := bytes.Index(content, sep)
+	if startIndex == -1 {
+		return "", errors.New("no gtid data in backup")
+	}
+	newOut := content[startIndex+len(sep):]
+	e := bytes.Index(newOut, []byte("'\n"))
+	if e == -1 {
+		return "", errors.New("can't find gtid data in backup")
+	}
+	content = content[:e]
+
+	se := bytes.Index(newOut, []byte("'"))
+	set := newOut[se+1 : e]
+
+	return string(set), nil
+}
+
+func (r *Recoverer) setBinlogs() error {
+	list, err := r.storage.ListObjects("binlog_")
+	if err != nil {
+		return errors.Wrapf(err, "list objects with prefix", "binlog_")
+	}
+	binlogs := []string{}
+	for _, binlog := range reverseArr(list) {
+		if strings.Contains(binlog, "-gtid-set") {
+			continue
+		}
+		binlogs = append(binlogs, binlog)
+
+		infoObj, err := r.storage.GetObject(binlog + "-gtid-set")
+		if err != nil {
+			return errors.Wrapf(err, "get %s object with gtid set", binlog)
+		}
+		content, err := ioutil.ReadAll(infoObj)
+		if err != nil {
+			return errors.Wrapf(err, "read %s gtid-set object", binlog)
+		}
+
+		isSubset, err := r.db.IsGTIDSubset(r.startGTID, string(content))
+		if err != nil {
+			return errors.Wrapf(err, "check if '%s' is a subset of '%s", r.startGTID, string(content))
+		}
+
+		if isSubset {
+			break
+		}
+	}
+	if len(binlogs) == 0 {
+		return errors.Errorf("no objects for prefix %s", "binlog_")
+	}
+	r.binlogs = reverseArr(binlogs)
+
+	return nil
+}
+
+func reverseArr(arr []string) []string {
+	for i := len(arr)/2 - 1; i >= 0; i-- {
+		opp := len(arr) - 1 - i
+		arr[i], arr[opp] = arr[opp], arr[i]
+	}
+
+	return arr
```

Comment:
just in case, there is no need in return: https://play.golang.org/p/lUeaDVuwmnS

design without return would be more aligned with std lib (see https://golang.org/pkg/sort/ for example)